icu4c/source/i18n/collationdatabuilder.cpp

```diff
@@ -1391,7 +1419,71 @@ CollationDataBuilder::buildMappings(CollationData &data, UErrorCode &errorCode)
+        // `CONTRACT_HAS_STARTER` flag set so that starters that
+        // can occur in a non-final position in a contraction can
+        // be easily recognized from having a contraction ce32
+        // that has the `CONTRACT_HAS_STARTER` flag set.
+
+        UCharsTrieBuilder contractionBuilder(errorCode);
+        // Intentionally unpaired low surrogate to make it never
+        // match well-formed UTF-16 which ICU4X feeds to the
+        // matcher.
+        UnicodeString placeholder(0xDC00);
+
+        UnicodeSetIterator iter(contractionMiddleStarter);
+        while(iter.next() && !iter.isString()) {
+            UChar32 c = iter.getCodepoint();
+            uint32_t ce32 = utrie2_get32(trie, c);
+            UBool fromBase = false;
+            if(ce32 == Collation::FALLBACK_CE32) {
+                fromBase = true;
+                ce32 = base->getCE32(c);
+            }
+            if (!(Collation::hasCE32Tag(ce32, Collation::CONTRACTION_TAG) && (ce32 & Collation::CONTRACT_HAS_STARTER))) {
+                if (fromBase) {
+                    // This case does not actually happen as of February 2025.
+                    ce32 = copyFromBaseCE32(c, ce32, true, errorCode);
+                }
+                if (Collation::hasCE32Tag(ce32, Collation::CONTRACTION_TAG)) {
+                    // This middle starter is also the first character of another
+                    // contraction, but that contraction does not have the
+                    // CONTRACT_HAS_STARTER flag. Let's add the flag to make
+                    // mark this at the expense of pessimizing the matching
+                    // of this contraction.
+                    // As of February 2025, this case does not actually occur
+                    // in CLDR.
+                    ce32 |= Collation::CONTRACT_HAS_STARTER;
+                } else {
+                    // This middle starter is not also the first character
+                    // in another contraction.
+
+                    // The UCharsTrie needs to contain some placeholder. Trying to
+                    // build an empty UCharsTrie causes an error. We build a trie
+                    // that never actually matches anything that ICU4X can try to
+                    // match, since ICU4X always passes well-formed UTF-16 to the
+                    // matcher and we put an unpaired low surrogate into the trie.
+                    // This pessimizes the character to CE mapping of the `c`,
+                    // since useless trie matching will be attempted but as of
+                    // February 2025, only two characters are affected.
+                    contractionBuilder.clear();
+                    contractionBuilder.add(placeholder, static_cast<int32_t>(ce32), errorCode);
+
+                    int32_t index = addContextTrie(ce32, contractionBuilder, errorCode);
+                    if(U_FAILURE(errorCode)) { return; }
+                    if(index > Collation::MAX_INDEX) {
```

Comment:
Does addContextTrie() not already check this?

Reply:
If it is checked, it's not obvious. Pre-existing calls to the method in this file had this check right after, which is why I copied the check here.